From 79878daca72e7a479fa182c8d84af10f4f7950f7 Mon Sep 17 00:00:00 2001
From: Pau Ruiz Safont <pau.safont@vates.tech>
Date: Fri, 12 Jun 2026 10:04:35 +0100
Subject: [PATCH] xapi_vm_snapshot: Fail VM.revert when a VDI.revert actually
 fails

In case where there's an error while reverting on the Storage Backend,
the state of the SR might become unstable, and a recovery using the
clone method for reverting might introduce bigger issues. Instead stop
the revert immediately.

The cases where ignoring the errors is acceptable is when either the SR
does not advertise support for VDI_REVERT, and when the SR does not
implement revert for a particular VDI. for example, if the SR implements
the revert for only some of the formats it implements.

Signed-off-by: Pau Ruiz Safont <pau.safont@vates.tech>
---
 ocaml/xapi/storage_smapiv1.ml  | 11 +++++++----
 ocaml/xapi/xapi_vdi.ml         |  3 +++
 ocaml/xapi/xapi_vm_snapshot.ml |  7 ++++++-
 3 files changed, 16 insertions(+), 5 deletions(-)

diff --git a/ocaml/xapi/storage_smapiv1.ml b/ocaml/xapi/storage_smapiv1.ml
index 3d39cf320..dcafdedea 100644
--- a/ocaml/xapi/storage_smapiv1.ml
+++ b/ocaml/xapi/storage_smapiv1.ml
@@ -1135,10 +1135,13 @@ module SMAPIv1 : Server_impl = struct
     let revert _context ~dbg ~sr ~snapshot_info =
       with_dbg ~name:"VDI.revert" ~dbg @@ fun di ->
       let dbg = Debug_info.to_string di in
-      Server_helpers.exec_with_new_task "VDI.revert"
-        ~subtask_of:(Ref.of_string dbg) (fun __context ->
-          call_revert ~__context ~dbg ~sr ~snapshot_info
-      )
+      try
+        Server_helpers.exec_with_new_task "VDI.revert"
+          ~subtask_of:(Ref.of_string dbg) (fun __context ->
+            call_revert ~__context ~dbg ~sr ~snapshot_info
+        )
+      with Smint.Not_implemented_in_backend ->
+        raise (Storage_error (Unimplemented "VDI.revert"))
   end
 
   let get_by_name _context ~dbg:_ ~name:_ = assert false
diff --git a/ocaml/xapi/xapi_vdi.ml b/ocaml/xapi/xapi_vdi.ml
index ab1e6063c..49235c717 100644
--- a/ocaml/xapi/xapi_vdi.ml
+++ b/ocaml/xapi/xapi_vdi.ml
@@ -1138,11 +1138,14 @@ let revert' ~__context ~snapshot =
   C.VDI.revert (Ref.string_of task) sr' snapshot_info
 
 let revert ~__context ~snapshot =
+  let __FUN = __FUNCTION__ in
   Storage_utils.transform_storage_exn @@ fun () ->
   try revert' ~__context ~snapshot
   with Storage_interface.Storage_error (Unimplemented _) ->
     debug "Backend does not implement VDI revert: doing it ourselves" ;
     let msg = [Ref.string_of (Db.VDI.get_SR ~__context ~self:snapshot)] in
+    debug "%s: Backend reported not implemented despite it offering the feature"
+      __FUN ;
     raise Api_errors.(Server_error (unimplemented_in_sm_backend, msg))
 
 let copy ~__context ~vdi ~sr ~base_vdi ~into_vdi =
diff --git a/ocaml/xapi/xapi_vm_snapshot.ml b/ocaml/xapi/xapi_vm_snapshot.ml
index 0f4324b27..95764a6c4 100644
--- a/ocaml/xapi/xapi_vm_snapshot.ml
+++ b/ocaml/xapi/xapi_vm_snapshot.ml
@@ -274,7 +274,12 @@ let revert_vbds ~__context ~rpc ~session_id ~snapshot ~vm =
         try
           Client.VDI.revert ~rpc ~session_id ~snapshot ;
           true
-        with _ -> false
+        with
+        | Api_errors.(Server_error (e, _))
+        when e = Api_errors.sr_operation_not_supported
+             || e = Api_errors.unimplemented_in_sm_backend
+        ->
+          false
       )
       snap_disks_all
   in
